map/honor-mobile-map-package-expiration-date/src/main/java/com/esri/samples/honor_mobile_map_package_expiration_date/HonorMobileMapPackageExpirationDateSample.java

```diff
@@ -0,0 +1,174 @@
+ * the License.
+ */
+
+package com.esri.samples.honor_mobile_map_package_expiration_date;
+
+import java.util.Calendar;
+import java.util.concurrent.TimeUnit;
+
+import javafx.animation.Animation;
+import javafx.animation.KeyFrame;
+import javafx.animation.Timeline;
+import javafx.application.Application;
+import javafx.geometry.Insets;
+import javafx.geometry.Pos;
+import javafx.scene.Scene;
+import javafx.scene.control.Alert;
+import javafx.scene.control.Label;
+import javafx.scene.layout.Background;
+import javafx.scene.layout.BackgroundFill;
+import javafx.scene.layout.CornerRadii;
+import javafx.scene.layout.StackPane;
+import javafx.scene.layout.VBox;
+import javafx.scene.paint.Paint;
+import javafx.scene.text.Font;
+import javafx.stage.Stage;
+import javafx.util.Duration;
+
+import com.esri.arcgisruntime.loadable.LoadStatus;
+import com.esri.arcgisruntime.mapping.Expiration;
+import com.esri.arcgisruntime.mapping.MobileMapPackage;
+import com.esri.arcgisruntime.mapping.view.MapView;
+
+public class HonorMobileMapPackageExpirationDateSample extends Application {
+
+  private MapView mapView;
+  private MobileMapPackage mobileMapPackage;
+
+  @Override
+  public void start(Stage stage) {
+
+    try {
+      // create stack pane and application scene
+      StackPane stackPane = new StackPane();
+      Scene scene = new Scene(stackPane);
+      scene.getStylesheets().add(getClass().getResource("/style.css").toExternalForm());
+
+      // set title, size, and add scene to stage
+      stage.setTitle("Honor Mobile Map Package Expiration Date Sample");
+      stage.setWidth(800);
+      stage.setHeight(700);
+      stage.setScene(scene);
+      stage.show();
+
+      // create an overlay to display the expiration information
+      VBox expirationOverlayVbox = new VBox(6);
+      expirationOverlayVbox.setBackground(new Background(new BackgroundFill(Paint.valueOf("rgba(0,0,0,0.3)"), CornerRadii.EMPTY,
+              Insets.EMPTY)));
+      expirationOverlayVbox.setPadding(new Insets(10.0));
+      expirationOverlayVbox.setMaxSize(800, 150);
+      expirationOverlayVbox.setAlignment(Pos.CENTER);
+      expirationOverlayVbox.getStyleClass().add("panel-region");
+
+      // create labels to display the expiration message and time since expiration
+      Label expirationMessageLabel = new Label();
+      expirationMessageLabel.setFont(new Font(16));
+      Label timeSinceExpirationLabel = new Label();
+      timeSinceExpirationLabel.setFont(new Font(16));
+
+      // add the labels to the overlay
+      expirationOverlayVbox.getChildren().addAll(expirationMessageLabel, timeSinceExpirationLabel);
+
+      // create a map view
+      mapView = new MapView();
+
+      // load the mobile map package
+      mobileMapPackage = new MobileMapPackage("./samples-data/mmpk/LothianRiversAnno.mmpk");
+      mobileMapPackage.loadAsync();
+      mobileMapPackage.addDoneLoadingListener(() -> {
```

Comment:
This is happening when we actually try and load the MMPK (line 133). I think that we can get the Expiration independently from loading it, so we don't actually have to do that check. @TADraeseke might be able to clarify?

Reply:
What I said yesterday may have been wrong--looks like the expiration info is only populated once the MMPK has been loaded or failed to load (https://master-dev.developers.arcgis.com/android/latest/api-reference/reference/com/esri/arcgisruntime/mapping/MobileMapPackage.html#getExpiration()). Could you check?

Reply:
So I've investigated, and it looks as if you must call `mmpk.loadAsync()`, but you don't actually have to verify the MMPK has `LoadStatus.LOADED` to do `mmpk.getExpiration()`. The reason for that would be that you can't load an MMPK that's expired (it will `LoadStatus.FAILED`), and might still want to get the `Expiration` to see the details. In which case I think it's sensible to do the checking of the load status afterwards, to demonstrate this point. What do others think?

Reply:
Hmmm what about:

```
    // wait for the map package to load
    mobileMapPackage.addDoneLoadingListener(() -> {
      // check if the map has expiration information and if so, has it expired yet
      if (mobileMapPackage.getExpiration() != null && mobileMapPackage.getExpiration().isExpired()) {
        // define a format for the date
        SimpleDateFormat daysHoursFormat = new SimpleDateFormat("yyyy-MM-dd' at 'hh:mm:ss", Locale.US);
        // show the expiration text view
        mExpirationMessageTextView.setVisibility(View.VISIBLE);
        // set the expiration message and expiration date to the text view
        mExpirationMessageTextView.setText(getString(R.string.expiration_text,
            mobileMapPackage.getExpiration().getMessage(),
            daysHoursFormat.format(mobileMapPackage.getExpiration().getDateTime().getTime())));
        if (mobileMapPackage.getExpiration().getType() == ExpirationType.ALLOW_EXPIRED_ACCESS) {
          // add the map to the map view
          mMapView.setMap(mobileMapPackage.getMaps().get(0));
        } else if (mobileMapPackage.getExpiration().getType() == ExpirationType.PREVENT_EXPIRED_ACCESS) {
          Toast.makeText(this, "The author of this mobile map package has disallowed access after the expiration data",
              Toast.LENGTH_LONG).show();
        }
      }
    });
    mobileMapPackage.loadAsync();
```

Set's the expiration message regardless of load status, sets the map to the mapview if allowed, informs the user if access is not allowed

Reply:
Looks good, would it still need a check for "LoadStatus.LOADED && !mmpk.getMaps().isEmpty" before we actually try and add it to the map view?

Reply:
When adding a map to the map view from an mmpk, we want to ensure this time that there is a map to add (so we need to check it's loaded) and also that there is a map in the mmpk to add (so checking that that .isEmpty is false)

Reply:
Yep, this has been added in recent commits. We check for expiration first (as Trevor suggested), to highlight this can be done without loading, then check for map availability and loading success before loading.